services/src/main/java/org/adamalang/services/social/Discord.java

```diff
@@ -22,46 +22,227 @@
+      if (emissionPossible) {
+        if (200 <= header.status && header.status <= 204) {
+          invokeSuccess = true;
+        } else {
+          emissionPossible = false;
+          logBody = HttpError.convert(header, logger,  ErrorCodes.WEB_STRING_CALLBACK_NOT_200, monitor, callback);
+        }
+      }
+    }
+
+    @Override
+    public void bodyStart(long size) {
+      if (size > 0) {
+        this.memory = new ByteArrayOutputStream((int) size);
+      }
+    }
+
+    @Override
+    public void bodyFragment(byte[] chunk, int offset, int len) {
+      if (memory == null) { // unknown size due to an error
+        this.memory = new ByteArrayOutputStream();
+      }
+      memory.write(chunk, offset, len);
+    }
+
+    @Override
+    public void bodyEnd() {
+      if (invokeSuccess && emissionPossible) {
+        String callbackVal = "";
+        if (memory != null) {
+          callbackVal = new String(memory.toByteArray(), StandardCharsets.UTF_8);
+        }
+        callback.success(callbackVal);
+        monitor.success();
+      }
+      if (logBody) {
+        if (memory != null) {
+          System.out.println(new String(memory.toByteArray(), StandardCharsets.UTF_8));
+        }
+        logger.error("failed body: {}", new String(memory.toByteArray(), StandardCharsets.UTF_8));
+      }
+    }
+
+    @Override
+    public void failure(ErrorCodeException ex) {
+      if (emissionPossible) {
+        emissionPossible = false;
+        logger.error("string-callback-failure:", ex);
+        monitor.failure(ex.code);
+        callback.failure(ex);
+      }
+    }
+  }
+  private void apiCall(String method, String[] uri, String[] requiredParams, TreeMap<String, String> headers, ObjectNode requestNode, String body, Callback<String> callback) {
+    for (String params : requiredParams) {
+      String hasParam = params;
+      if (params.startsWith("_")) {
+        hasParam = params.substring(1);
+      }
+      if (!requestNode.has(hasParam)) {
+        callback.failure(new ErrorCodeException(888890, "Method is missing parameter: " + hasParam));
+        return;
+      }
+    }
+    StringBuilder uriBuilder = new StringBuilder();
+    uriBuilder.append("https://discord.com/api");
+    int pos = 0;
+    while (pos < (requiredParams.length)) {
+      if (requiredParams[pos].startsWith("_")) {
+        break;
+      }
+      pos++;
+      uriBuilder.append(uri[pos - 1]).append("/" + requestNode.get(requiredParams[pos - 1]).textValue());
+    }
+    for (int i = pos; i < uri.length; i++) {
+      uriBuilder.append(uri[i]);
+    }
+    SimpleHttpRequestBody requestBody;
+    if (method.toUpperCase() == "GET" || method.toUpperCase() == "DELETE") {
+      requestBody = SimpleHttpRequestBody.EMPTY;
+    } else {
+      requestBody = SimpleHttpRequestBody.WRAP(body.getBytes(StandardCharsets.UTF_8));
+    }
+    SimpleHttpRequest req = new SimpleHttpRequest(method, uriBuilder.toString(), headers, requestBody);
```

Comment:
in light of the prior comment, since the construction SimpleHttpRequest is predictable, it would be great to have a DiscordSimpleHttpRequestFactory along with unit tests and coverage. The unit tests in this case are a way of encoding expectations for the next person around behavior.

Reply:
I am a bit confused. If I were to create unit tests, would that just be a way to show the form and expectations of a request for Discord?

Reply:
Yes, the unit test in this case gives a way to validate that the Discord specific inputs are converted appropriately to a SimpleHttpRequest whether you call this class DiscordRequestBuilder or DiscordSimpleHttpRequestFactory